File: clients/php/src/Hasher.php
<?php

namespace Vexilla;

class Hasher
{
    private static $FNV_OFFSET_BASIS = 2166136261;
    private static $FNV_PRIME = 16777619;

    /**
     * @param float $stringToHash
     */
    static function hashString(string $stringToHash, float $seed)
    {
        $splitString = str_split($stringToHash);
        $length = count($splitString);
        $hashResult = self::$FNV_OFFSET_BASIS;

        for ($i = 0; $i < $length; $i++) {
            $hashResult = abs($hashResult ^ mb_ord($splitString[$i]));
            $hashResult = ($hashResult * self::$FNV_PRIME) & 0xFFFFFFFF;
        }

        return ($hashResult * $seed) % 1000 / 1000;
    }
}

File: clients/kotlin/src/main/kotlin/Hasher.kt
package dev.vexilla

import kotlin.math.floor

object Hasher {

    // private var manifest: Manifest? = null
    // private const uint FNV32_OFFSET_BASIS = 2166136261;
    // private const uint FNV32_PRIME = 16777619;

    private var FNV32_OFFSET_BASIS: UInt = 2166136261u
    private var FNV32_PRIME: UInt = 16777619u

    fun hashString(stringToHash: String, seed: Double): Double {
        var hashValue = FNV32_OFFSET_BASIS

        val length = stringToHash.length - 1
        for (i in 0..length) {
            val char = stringToHash[i]
            hashValue = hashValue xor char.code.toUInt()
            hashValue *= FNV32_PRIME
        }

        return floor(hashValue.toDouble() * seed * 42) % 1000 / 1000
    }

    fun hashInt(numberToHash: Int, seed: Double): Double {
        return Hasher.hashString("$numberToHash", seed)
    }

    fun hashLong(numberToHash: Long, seed: Double): Double {
        return Hasher.hashString("$numberToHash", seed)
    }

    fun hashFloat(numberToHash: Float, seed: Double): Double {
        return Hasher.hashString("$numberToHash", seed)
    }

    fun hashDouble(numberToHash: Double, seed: Double): Double {
        return Hasher.hashString("$numberToHash", seed)
    }

}
